Support module imports using a qualified name

hubertp opened the pull request:
### Pull Request Description

This change allows for importing modules using a qualified name and deals with any conflicts on the way.
Given a module C defined at `A/B/C.enso` with
```
type C
    type C a
```
it is now possible to import it as
```
import project.A
...
    val x = A.B.C 10
```

Given a module located at `A/B/C/D.enso`, we will generate
intermediate, synthetic, modules that only import and export the successor module along the path.
For example, the contents of a synthetic module B will look like
```
import <namespace>.<pkg-name>.A.B.C
export <namespace>.<pkg-name>.A.B.C
```
If module B is defined already by the developer, the compiler will _inject_ the above statements to the IR.

Also removed the last elements of some lowercase name resolution that managed to survive recent
changes (`Meta.Enso_Project` would now be ambiguous with `enso_project` method).

Finally, added a pass that detects shadowing of the synthetic module by the type defined along the path.
We print a warning in such a situation.

Related to https://www.pivotaltracker.com/n/projects/2539304

### Important Notes

There was an additional request to fix the annoying problem with `from` imports that would always bring 
the module into the scope. The changes in stdlib demonstrate how it is now possible to avoid the workaround of
```
from X.Y.Z as Z_Module import A, B
```
(i.e. `as Z_Module` part is almost always unnecessary).

### Checklist

Please include the following checklist in your PR:

- [ ] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide dist` and `./run ide watch`.

hubertp committed:
Allow for importing submodules

Easy path for importing submodules.
Given a module located at `A/B/C/D.enso`, we will generate
intermediate, virtual, modules that only import and export the succesor.
For example, the contents of a virtual module B will look like
```
import <namespace>.<pkg-name>.A.B.C
export <namespace>.<pkg-name>.A.B.C
```

This change does not attempt to deal with conflicts, yet.

hubertp committed:
Deal with conflicts

When a virtual module conflicts with a non-virtual module, magic
happens.
We enhance the non-virtual module with essentially exactly the same
information that the virtual one would carry without loosing anything.

In order to distinguish between regular imports and the virtual ones, an
additional flag needs to be carried in order to be able to continue to
do name resolution properly.

Also fixed a long standing problem with imports that include names but
also import the name of the module automatically. Binding map was
incorrectly always taking into account the name of the import.

hubertp committed:
Remove workarounds for import problems

Given
```
from X.Y.Z import A, B
```
would always bring `Z` into scope as well which was almost never
desirable. As a result stdlib was full of
```
from X.Y.Z as Z_Module import A, B
```
workarounds.

This is now no longer necessary.

hubertp committed:
nit

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:
```diff
@@ -154,15 +160,18 @@ public Module(QualifiedName name, Package<TruffleFile> pkg, Rope literalSource)
-  private Module(QualifiedName name, Package<TruffleFile> pkg) {
-    this.sources = ModuleSources.NONE;
+  private Module(
+      QualifiedName name, Package<TruffleFile> pkg, boolean virtual, Optional<Rope> literalSource) {
```
My preference in Java is to use `Rope literalSource` and then check for `null`.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:
```diff
@@ -154,15 +160,18 @@ public Module(QualifiedName name, Package<TruffleFile> pkg, Rope literalSource)
     this.scope = new ModuleScope(this);
     this.pkg = pkg;
-    this.compilationStage = CompilationStage.AFTER_CODEGEN;
+    this.compilationStage = virtual ? CompilationStage.INITIAL : CompilationStage.AFTER_CODEGEN;
```
Virtual modules, that's something @4e6 may also be interested in for his _lazy visualizations_.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:
```diff
@@ -198,6 +216,27 @@ public void setLiteralSource(String source) {
+   *
+   * @param names fully qualified names of (potentially virtual) modules' names
+   */
+  public void setDirectSubmoduleNames(QualifiedName[] names) {
```
How can we make sure this value doesn't change more than once? It shouldn't change, right?

hubertp replied:
It shouldn't, added an assert that will blow it up if someone changes the logic.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:
```diff
@@ -198,6 +216,27 @@ public void setLiteralSource(String source) {
+    if (directSubmoduleNames == null) {
+      return new QualifiedName[0];
+    }
+    return directSubmoduleNames;
```
Beware the array is modifiable! Returning `Collections.unmodifiableList(Arrays.asList(directSubmoduleNames))` would be safer.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/PackageRepository.scala:
```diff
@@ -457,8 +533,31 @@ object PackageRepository {
+      } else {
+        val loaded = loadedModules(virtualModule.getName.toString)
+        if (!loaded.isVirtual && virtualModule.isVirtual) {
+          loaded.setDirectSubmoduleNames(refs.toArray)
```
Maybe easier to convert `refs` to Java (unmodifiable) `List` rather than array.

hubertp replied:
:+1:

JaroslavTulach commented on engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/ImportsTest.scala:
```diff
@@ -77,4 +77,13 @@ class ImportsTest extends PackageTest {
   }
+
+  "Import statements" should "should allow for importing submodules" in {
+    evalTestProject("TestSubmodules") shouldEqual 42
```
Is the `TestSubmodules` project evaluated with standard distribution or micro distribution?

hubertp replied:
It doesn't matter TBH. I only import from stdlib println because I want to do additional verification and not deal with unused variable warnings. Strictly speaking it is not necessary because compiler not complaining is already a progress.

JaroslavTulach reviewed (approved):
**Request**: Make sure the submodules setter can be invoked only once and the value cannot change when the system is running.

_Advice_: Make sure the submodules getter returns immutable array/`List`.
_Advice_: If the submodules setter can only be invoked by `IR` package, consider using [friend packages](http://wiki.apidesign.org/wiki/FriendPackages) pattern.

_Opinion_: Avoid `Optional` in Java and check for `null`

JaroslavTulach commented:
Btw. the meaning of

> **Request**, _Advice_, _Opinion_

is taken from SunARC (Sun Microsystems Architecture Review Committee) guidelines as [adopted by NetBeans](http://wiki.apidesign.org/wiki/APIReviewSteps). I am approving only under the assumption that all **Requests** are addressed. I'd be glad to see _Advices_ implemented - but you can explain why that's not good idea and I'll live with that. _Opinions_ reflect my personal preferences and you can ignore them.

farmaazon reviewed (approved):
LGTM

hubertp committed:
Address PR review

jdunkerley commented on distribution/lib/Standard/Base/0.0.0-dev/src/System/File.enso:
```diff
@@ -6,13 +6,11 @@ import Standard.Base.System.File.File_Permissions
 import Standard.Base.Data.Text.Text_Sub_Range
-from Standard.Base.Data.Text.Encoding as Encoding_Module import Encoding, Encoding_Error
-from Standard.Base.Error.Problem_Behavior as Problem_Behavior_Module import Problem_Behavior, Report_Warning
+from Standard.Base.Data.Text.Encoding import Encoding, Encoding_Error
+from Standard.Base.Error.Problem_Behavior import Problem_Behavior, Report_Warning
```
```suggestion
from Standard.Base.Error.Problem_Behavior import Report_Warning
```

Others are the `from Standard.Base import all`

jdunkerley commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso:
```diff
@@ -8,10 +8,10 @@ import Standard.Base.Data.Text.Matching_Mode
 import Standard.Base.Data.Text.Text_Sub_Range
-from Standard.Base.Data.Text.Encoding as Encoding_Module import Encoding, Encoding_Error
-from Standard.Base.Error.Problem_Behavior as Problem_Behavior_Module import Problem_Behavior, Report_Warning
+from Standard.Base.Data.Text.Encoding import Encoding, Encoding_Error
+from Standard.Base.Error.Problem_Behavior import Problem_Behavior, Report_Warning
```
```suggestion
from Standard.Base.Error.Problem_Behavior import Report_Warning
```
Others in the standard import.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:
```diff
@@ -221,20 +222,21 @@ public void setLiteralSource(String source) {
-      return new QualifiedName[0];
+  public List<QualifiedName> getDirectVirtualModulesRefs() {
+    if (directVirtualModulesRefs == null) {
+      return new ArrayList<>();
```
`Collections.emptyList()` would be better.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:
```diff
@@ -221,20 +222,21 @@ public void setLiteralSource(String source) {
-  public void setDirectSubmoduleNames(QualifiedName[] names) {
-    directSubmoduleNames = names;
+  public void setDirectVirtualModulesRefs(List<QualifiedName> names) {
+    assert directVirtualModulesRefs == null;
```
OK.

hubertp committed:
nit

jdunkerley reviewed (approved):
Looks good to me.

radeusgd commented on engine/runtime/src/main/scala/org/enso/compiler/data/BindingsMap.scala:
```diff
@@ -727,6 +742,14 @@ object BindingsMap {
+      * @return true, if generated by the compiler, false otherwise
+      */
+    def isVirtual(): Boolean = {
+      importDef.isVirtual && exports.map(_.isVirtual).getOrElse(false)
```
It is slightly unclear to me why this is an `&&` and not an `||`? In which case can it happen that the `importDef` would be virtual but exports not?

hubertp replied:
Both should always be virtual because both are added by the compiler at the same time.

radeusgd replied:
That's what I expected and I think that's why it got me confused why we have `&&`.
Maybe we should do 
```suggestion
      if (importDef.isVirtual) {
         assert exports.map(_.isVirtual).contains(true)
         true
      } else false
```

to express that it is an 'iff' relation?

Or ignore the assert but only check the import if we are sure export is going to be the same?
```suggestion
      importDef.isVirtual
```

The current code checks for the exports and if they were non-virtual it will return false even for a virtual import. As this situation should never happen in practice, IMO the secondary check for exports is redundant. If we want to keep it, I'd suggest to at least report (via assertion or internal compiler error) that an illegal state has been encountered.

hubertp replied:
Fair enough, I will remove the redundant part.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Ordering/Comparator.enso:
```diff
@@ -1,6 +1,6 @@
 from Standard.Base import all
 import Standard.Base.Data.Ordering.Natural_Order
-from Standard.Base.Data.Text.Text_Ordering as Text_Ordering_Module import Text_Ordering
+from Standard.Base.Data.Text.Text_Ordering import Text_Ordering
```
I just wanted to say how awesome that change is! The need for this renaming in so many places was really quite annoying.

hubertp replied:
![image](https://user-images.githubusercontent.com/292128/181556039-1870fbf2-55ba-40af-b46f-777fbc4e4868.png)

radeusgd commented on engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/ImportsTest.scala:
```diff
@@ -77,4 +77,13 @@ class ImportsTest extends PackageTest {
     consumeOut should contain("Export statements form a cycle:")
   }
+
+  "Import statements" should "should allow for importing submodules" in {
```
I remember discussions on what to do when we both have stuff like module `A/B.enso` defining `type C` and a module `A/B/C.enso`.

I don't see any tests checking how this conflict-case behaves. Can we add those? (Or did I miss something?)

hubertp replied:
Yeap, was going to add it. Will need to add some additional check to produce a reasonable message

hubertp replied:
@radeusgd done

radeusgd reviewed (approved):
Looks good, I really appreciate not having to rename modules when it's not truly necesary.

I think a test for conflicting module names should be added. We should also probably ensure that if any error message is emitted due to that event, it is quite clear to the user what is the issue and how it can be solved.

hubertp committed:
Linting pass for detecting type/module conflicts

Made it a warning, since this only manifests itself when one tries to
access the conflicting module.

hubertp committed:
Apply suggestions from code review

Co-authored-by: James Dunkerley <jdunkerley@users.noreply.github.com>

hubertp committed:
Update engine/runtime/src/main/scala/org/enso/compiler/data/BindingsMap.scala

Co-authored-by: Radosław Waśko <radoslaw.wasko@enso.org>

hubertp committed:
changelog update

hubertp committed:
documentation and minor code cleanups

kustosz commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java:
```diff
@@ -109,6 +113,7 @@ public Module(QualifiedName name, Package<TruffleFile> pkg, TruffleFile sourceFi
     this.cache = new ModuleCache(this);
     this.wasLoadedFromCache = false;
     this.hasCrossModuleLinks = false;
+    this.virtual = false;
```
I think you said you prefer `synthetic` in some discord discussion? I'm with you, it's probably a better name, given `virtual` means something very different in PLD.

hubertp replied:
To paraphrase the known quote, naming things is one of the two hard things in CS....
I will make the change. Thanks for taking a look!

hubertp committed:
s/virtual/synthetic

hubertp committed:
Make windows happy
